docs/sdk/instructlab-sdk.md

```diff
@@ -0,0 +1,109 @@
+
+Unifying these various ranges of advanced workflows under an overarching `InstructLab SDK API` will allow for new usage patterns and a clearer story on what InstructLab can and should provide as user accessible endpoints.
+
+While each library can and _should_ have their own publicly accessible API, not all functionality being added to SDG, Training, and Eval needs to be correlated directly to the "InstructLab workflow". This Python SDK should, as the CLI does, expose an opinionated flow that uses functionality from the various libraries. The InstructLab SDK should be derived from the library APIs, not the other way around. SDG, for example, currently has a `generate_data` method, meant to only be accessed by InstructLab. This method simply calls other publicly available SDG functionality. Orchestration of the InstructLab flow like this should not be of concern to the individual libraries and instead be handled by the overarching InstructLab SDK which will maintain the user contracts. The InstructLab SDK will need to work within the bounds of what the Libraries expose as public APIs.
+
+The benefit of the above is that the opinionated flow can be accessed in a more nuanced and piecemeal way while also gaining the potential for more advanced features. Say a consumer wants to:
+
+1. setup a custom config file for ilab (optional)
+2. initialize a taxonomy
+3. ensure their taxonomy is valid
+4. ingest some data for RAG and SDG (SDG coming soon)
+5. generate synthetic data using an InstructLab pipeline
+6. do some custom handling per their use case
+7. fine-tune a model using the custom config they initialized for their hardware
+
+a user could do the following to accomplish this if they had an SDK:
+
+```python
+instructlab.api.configuration.init(taxonomy=path_to_taxonomy)
+if instructlab.api.taxonomy.diff():
+    instructlab.api.rag.ingest(...)
+    instructlab.api.data.ingest(...)
+    data_paths = instructlab.api.data.generate(...)
+    some_custom_handling(data_paths)
+    instructlab.api.model.train(data_path=..., strategy=lab_multiphase)
+```
+
+However today, users are forced to run a sequence of commands tailored to only work with the proper directory structure on the system.
+
+## Major Goals
+
+1. Modularize the InstructLab workflow such that any part can be run independently
+2. Allow users to choose whether or not to take advantage of the config/system-profile method of running InstructLab. Meaning they do not need any pre-existing configuration to run the SDK.
+3. Standardize user contracts for the existing functionality of the InstructLab workflow. Existing CLI commands should be using the SDK once past click parsing, not separate code.
+4. Define Contracts loose enough that functionality can be expanded as more advanced features are released.
+5. Document SDK usage in upcoming InstructLab releases.
+
+## Non-Goals
```

Comment:
I love that you've included this as part of the dev-doc.

Reply:
thanks!